// @chatbots-studio/cbs-platform-node-red-palette/helpers/createTemplate.js
const axios = require("axios");
const urlJoin = require("url-join");

const {
  envConfig: { messageApiEnvConfig },
} = require("../config");
const { handleNodeError } = require("../error");
const { customLogger } = require("../logger");

module.exports = async function (
  { description, group, json, templateId, timer, type },
  node
) {
  const logger = customLogger(node);

  try {
    const env = await messageApiEnvConfig();

    const template = {
      channel: env.CHANNEL,
      companyId: env.COMPANY_ID,
      botId: env.BOT_ID,
      description,
      group,
      json: JSON.parse(json),
      templateId,
      timer,
      type,
    };

    const requestOptions = {
      method: "POST",
      headers: {
        authorization: env.MESSAGES_API_KEY,
        "content-type": "application/json",
      },
      url: urlJoin(env.MESSAGES_API_URL, "v1/templates"),
      data: template,
    };

    logger.addParam("outUrl", requestOptions.url);
    logger.addParam("request", requestOptions);

    const response = await axios(requestOptions);

    logger.addParam("statusCode", response.status);

    logger.addParam("response", response.data);

    logger.logInfo();
    return response.data;
  } catch (error) {
    await handleNodeError(error, logger, node);

    return { error };
  }
};
