// @chatbots-studio/cbs-platform-node-red-palette/enums/exampleJsons.js
module.exports = {
  TG: {
    text: {
      helpLink: 'https://core.telegram.org/bots/api#message',
      json: {
        "text": "{{text}}",
        "en": {
          "text": "hello message"
        }
      }
    },
    textAndKeyboard: {
      helpLink: 'https://core.telegram.org/bots/api#replykeyboardmarkup',
      json: {
        "text": "{{text}}",
        "parse_mode": "Markdown",
        "keyboard": [
          [
            {
              "text": "button1"
            },
            {
              "text": "button2"
            }
          ],
          [
            {
              "text": "button3"
            },
            {
              "text": "button4"
            }
          ]
        ],
        "en": {
          "text": "hello message",
          "button1": "Hello",
          "button2": "Hi",
          "button3": "Hola",
          "button4": "Hey"
        },
        "otk": true
      }
    },
    textAndCallbackButtons: {
      helpLink: 'https://core.telegram.org/bots/api#inlinekeyboardbutton',
      json: {
        "text": "{{text}}",
        "inline": [
          [
            {
              "callback_data": "{{callback1}}",
              "text": "{{callbackText1}}"
            }
          ],
          [
            {
              "callback_data": "{{callback2}}",
              "text": "{{callbackText2}}"
            }
          ]
        ],
        "en": {
          "text": "hello message",
          "callback1": "menu",
          "callback2": "back",
          "callbackText1": "Menu",
          "callbackText2": "Back"
        }

      }
    },
    photoAndTextAndCallbackButtons: {
      helpLink: 'https://core.telegram.org/bots/api#inlinekeyboardbutton',
      json: {
        "type": "photo",
        "photo": "{{picture}}",
        "caption": "{{text}}",
        "inline": [
          [
            {
              "callback_data": "{{url}}",
              "text": "{{callbackText1}}"
            }
          ],
          [
            {
              "callback_data": "{{callback2}}",
              "text": "{{callbackText2}}"
            }
          ]
        ],
        "en": {
          "text": "hello message",
          "callback1": "https://core.telegram.org/bots/api#inlinekeyboardmarkup",
          "callback2": "back",
          "callbackText1": "Menu",
          "callbackText2": "Back"
        }
      }
    },
    photoAndText: {
      helpLink: 'https://core.telegram.org/bots/api#message',
      json: {
        "type": "photo",
        "photo": "{{cardPic}}",
        "caption": "{{text}}",
        "en": {
          "text": "hello message"
        }
      }
    },
    textAndRemoveKeyboard: {
      helpLink: 'https://core.telegram.org/bots/api#replykeyboardremove',
      json: {
        "text": "{{text}}",
        "remove_keyboard": true,
        "en": {
          "text": "hello message"
        }
      },
    },
    requestPhoneAndText: {
      helpLink: 'https://core.telegram.org/bots/api#keyboardbutton',
      json: {
        "text": "{{text}}",
        "keyboard": [
          [
            {
              "request_contact": true,
              "text": "{{button}}"
            }
          ]
        ],
        "en": {
          "text": "share phone",
          "button1": "phone"
        }
      },
    },
    photoAndTextAndKeyboard: {
      helpLink: 'https://core.telegram.org/bots/api#keyboardbutton',
      json: {
        "type": "photo",
        "photo": "{{picture}}",
        "caption": "{{text}}",
        "keyboard": [
          [
            {
              "text": "{{buttonText1}}"
            },
            {
              "text": "{{buttonText2}}"
            }
          ],
          [
            {
              "text": "{{buttonText3}}"
            },
            {
              "text": "{{buttonText4}}"
            }
          ],
          [
            {
              "text": "{{buttonText5}}"
            },
            {
              "text": "{{buttonText6}}"
            }
          ],
          [
            {
              "text": "{{buttonText7}}"
            },
            {
              "text": "{{buttonText8}}"
            }
          ]
        ],
        "ru": {
          "text": "hello message:",
          "buttonText1": "button1",
          "buttonText2": "button2",
          "buttonText3": "button3",
          "buttonText4": "button4",
          "buttonText5": "button5",
          "buttonText6": "button6",
          "buttonText7": "button7",
          "buttonText8": "button8"
        },
        "otk": true
      },
    }
  }
}
