// @chatbots-studio/cbs-platform-node-red-palette/config/config.js
module.exports = {
  customerApiEnvConfig: async () => {
    const config = {
      CUSTOMER_API_KEY: process.env.CUSTOMER_API_KEY,
      CUSTOMER_API_URL: process.env.CUSTOMER_API_URL,
    };

    await Object.entries(config).forEach(([key, value]) => {
      if (!value) {
        throw new Error(`${ key } not configured in Environment Variables`);
      }
    });

    return config;
  },
  messageApiEnvConfig: async () => {
    const config = {
      BOT_ID: process.env.BOT_ID,
      CHANNEL: process.env.CHANNEL,
      COMPANY_ID: process.env.COMPANY_ID,

      MESSAGES_API_KEY: process.env.MESSAGES_API_KEY,
      MESSAGES_API_URL: process.env.MESSAGES_API_URL,
    };

    // await Object.entries(config).forEach(([key, value]) => {
    //   if (!value) {
    //     throw new Error(`${ key } not configured in Environment Variables`);
    //   }
    // });
    if (!config.MESSAGES_API_KEY || !config.MESSAGES_API_URL) {
      throw new Error(`MESSAGES_API_KEY or MESSAGES_API_URL not configured in Environment Variables`);
    }

    return config;
  }
};
